File: cko-projeto/cko-projeto/cko-projeto/03-templates/ckos_runtime.py
#!/usr/bin/env python3
"""
CKOS Runtime Engine — implementação de referência do ckosRuntimeLayer (CKO Seringa v11).

Executa o pipeline declarado em `ckosRuntimeLayer.stages` sobre um documento CKO e um
contexto de paciente, aplicando guardrails de segurança e emitindo um executionTrace
auditável. É a "camada de execução" que transforma os engines estáticos do schema em
um sistema que processa uma consulta de ponta a ponta.

Fronteira de escopo (honesta, sem placeholders):
  - AVALIA a forma executável das regras (conditions/actions) em ruleRepositories e em
    clinicalKnowledge.clinicalRules.
  - NÃO interpreta `governedRule.logic` (string/DSL) — essa é a forma de governança;
    a forma executável é conditions/actions. Um parser de DSL é trabalho de v12.
  - As shaclShapes de knowledgeGraph.semanticWeb são declaradas, não aplicadas aqui.
  - Validação estrutural do documento usa jsonschema (opcional) quando disponível.

Uso:
    python3 ckos_runtime.py --schema seringa-cko-v11.schema.json \
                            --cko seringa-insulina-030.cko.json \
                            --patient '{"medication":"insulin","volumeMl":0.3,...}'
"""
from __future__ import annotations
import argparse
import datetime as _dt
import json
import uuid
from dataclasses import dataclass, field
from typing import Any


# --------------------------------------------------------------------------- #
# Utilidades
# --------------------------------------------------------------------------- #
def _now() -> str:
    return _dt.datetime.now(_dt.timezone.utc).isoformat()


def _get(ctx: dict, key: str) -> Any:
    """Leitura de campo do contexto, tolerante a subestruturas {value,unit}."""
    val = ctx.get(key)
    if isinstance(val, dict) and "value" in val:
        return val["value"]
    return val


# --------------------------------------------------------------------------- #
# Avaliador de condições / regras
# --------------------------------------------------------------------------- #
_OPS = {
    "equals":        lambda a, b: a == b,
    "notEquals":     lambda a, b: a != b,
    "greaterThan":   lambda a, b: a is not None and a > b,
    "lessThan":      lambda a, b: a is not None and a < b,
    "greaterOrEqual":lambda a, b: a is not None and a >= b,
    "lessOrEqual":   lambda a, b: a is not None and a <= b,
    "in":            lambda a, b: a in b if isinstance(b, (list, tuple, set)) else False,
    "notIn":         lambda a, b: a not in b if isinstance(b, (list, tuple, set)) else True,
    "exists":        lambda a, b: (a is not None) == bool(b),
}


def eval_condition(cond: dict, ctx: dict) -> bool:
    op = _OPS.get(cond["operator"])
    if op is None:
        raise ValueError(f"Operador desconhecido: {cond['operator']}")
    return bool(op(_get(ctx, cond["field"]), cond.get("value")))


def eval_rule(rule: dict, ctx: dict) -> bool:
    """Regra dispara se TODAS as condições forem verdadeiras (AND)."""
    conds = rule.get("conditions") or []
    if not conds:
        return False
    return all(eval_condition(c, ctx) for c in conds)


# --------------------------------------------------------------------------- #
# Trace de execução
# --------------------------------------------------------------------------- #
@dataclass
class ExecutionTrace:
    traceId: str = field(default_factory=lambda: f"trace-{uuid.uuid4()}")
    startedAt: str = field(default_factory=_now)
    completedAt: str | None = None
    stagesExecuted: list = field(default_factory=list)
    firedRules: list = field(default_factory=list)
    assertedFacts: list = field(default_factory=list)
    decisionPath: list = field(default_factory=list)
    finalRecommendation: dict | None = None
    warnings: list = field(default_factory=list)
    safetyBlocks: list = field(default_factory=list)
    confidence: float | None = None
    humanValidated: bool = False
    status: str = "pending"          # pending | approved | pending_human_validation | blocked | error
    reasoningTrace: list = field(default_factory=list)

    def to_dict(self) -> dict:
        d = self.__dict__.copy()
        return d


# --------------------------------------------------------------------------- #
# Runtime
# --------------------------------------------------------------------------- #
class CKOSRuntime:
    def __init__(self, cko: dict):
        self.cko = cko
        self.layer = cko.get("ckosRuntimeLayer", {})
        self.trace = ExecutionTrace()

    # ---- validação estrutural opcional ---------------------------------- #
    @staticmethod
    def validate(cko: dict, schema: dict) -> list[str]:
        try:
            from jsonschema import Draft202012Validator
        except ImportError:
            return ["jsonschema não instalado — validação estrutural pulada."]
        v = Draft202012Validator(schema)
        return [f"{list(e.path)}: {e.message}" for e in v.iter_errors(cko)]

    # ---- handlers de estágio -------------------------------------------- #
    def _stage_ingest(self, ctx: dict, stage: dict):
        self.trace.reasoningTrace.append({"stage": "ingest", "detail": f"{len(ctx)} campos de contexto carregados."})

    def _stage_fact_assertion(self, ctx: dict, stage: dict):
        for a in self.cko.get("factInferenceEngine", {}).get("assertions", []):
            self.trace.assertedFacts.append({"id": a["id"], "fact": a["fact"], "confidence": a.get("confidence", 1.0)})
        self.trace.reasoningTrace.append(
            {"stage": "factAssertion", "detail": f"{len(self.trace.assertedFacts)} fatos assertidos da base."})

    def _stage_rule_evaluation(self, ctx: dict, stage: dict):
        repos = self.cko.get("ruleRepositories", {})
        # também avalia regras inline de clinicalKnowledge
        inline = self.cko.get("clinicalKnowledge", {}).get("clinicalRules", [])
        buckets = [("clinicalRules", repos.get("clinicalRules", []) + inline),
                   ("safetyRules", repos.get("safetyRules", [])),
                   ("inventoryRules", repos.get("inventoryRules", [])),
                   ("educationRules", repos.get("educationRules", []))]
        for domain, rules in buckets:
            for r in sorted(rules, key=lambda x: x.get("priority", 100)):
                if eval_rule(r, ctx):
                    fired = {"id": r.get("id", r.get("name", "unnamed")),
                             "domain": r.get("domain", domain),
                             "safetyCritical": bool(r.get("safetyCritical", False)),
                             "confidence": r.get("confidence", 0.8),
                             "actions": r.get("actions", [])}
                    self.trace.firedRules.append(fired)
                    for act in r.get("actions", []):
                        if act["type"] in ("avoid", "block") and fired["safetyCritical"]:
                            self.trace.safetyBlocks.append({"rule": fired["id"], "target": act["target"],
                                                            "message": act.get("message", "")})
        self.trace.reasoningTrace.append(
            {"stage": "ruleEvaluation",
             "detail": f"{len(self.trace.firedRules)} regras dispararam; {len(self.trace.safetyBlocks)} bloqueios de segurança."})

    def _stage_inference(self, ctx: dict, stage: dict):
        known = {f["id"] for f in self.trace.assertedFacts}
        changed = True
        rules = self.cko.get("factInferenceEngine", {}).get("inferenceRules", [])
        while changed:  # forward chaining simples
            changed = False
            for r in rules:
                antecedents = set(r.get("if", []))
                if antecedents and antecedents <= known:
                    for c in r.get("then", []):
                        if c not in known:
                            known.add(c)
                            self.trace.assertedFacts.append({"id": c, "fact": c, "confidence": 0.9, "inferred": True})
                            changed = True
        self.trace.reasoningTrace.append(
            {"stage": "inference", "detail": f"{len([f for f in self.trace.assertedFacts if f.get('inferred')])} fatos inferidos."})

    def _stage_decision(self, ctx: dict, stage: dict):
        # 1) árvore de decisão, se houver
        for tree in self.cko.get("decisionTreeRepository", {}).get("trees", []):
            nodes = {n["id"]: n for n in tree.get("nodes", [])}
            cur = tree.get("rootNode")
            while cur and cur in nodes:
                node = nodes[cur]
                self.trace.decisionPath.append({"node": cur, "question": node.get("question")})
                if node.get("type") in ("terminal", "action"):
                    break
                answer = eval_condition(node["condition"], ctx) if "condition" in node else _get(ctx, node.get("field", ""))
                cur = node.get("yes") if answer else node.get("no")
        # 2) recomendação a partir de regras de ação 'recommend'/'require'
        recs = [a for fr in self.trace.firedRules for a in fr["actions"] if a["type"] in ("recommend", "require")]
        if recs:
            self.trace.finalRecommendation = {"recommendedDevice": recs[0]["target"],
                                              "actionType": recs[0]["type"],
                                              "supportingRules": [fr["id"] for fr in self.trace.firedRules
                                                                  if any(a in recs for a in fr["actions"])]}
        self.trace.reasoningTrace.append(
            {"stage": "decision",
             "detail": f"decisão={self.trace.finalRecommendation}; caminho de {len(self.trace.decisionPath)} nós."})

    def _stage_workflow(self, ctx: dict, stage: dict):
        wfs = self.cko.get("clinicalWorkflowEngine", {}).get("workflows", [])
        self.trace.reasoningTrace.append({"stage": "workflow", "detail": f"{len(wfs)} workflow(s) associado(s)."})

    def _stage_explanation(self, ctx: dict, stage: dict):
        # confiança = menor confiança entre as regras que sustentam a recomendação
        supporting = self.trace.finalRecommendation.get("supportingRules", []) if self.trace.finalRecommendation else []
        confs = [fr["confidence"] for fr in self.trace.firedRules if fr["id"] in supporting] or \
                [fr["confidence"] for fr in self.trace.firedRules] or [0.0]
        self.trace.confidence = round(min(confs), 3)
        self.trace.reasoningTrace.append(
            {"stage": "explanation",
             "detail": f"confiança agregada={self.trace.confidence} (mínimo das regras de suporte)."})

    def _stage_human_validation(self, ctx: dict, stage: dict):
        g = self.layer.get("globalGuardrails", {})
        sg = stage.get("guardrails", {})
        needs_human = False
        reasons = []
        # 1) bloqueio de segurança nunca passa sem humano
        if self.trace.safetyBlocks:
            needs_human = True
            reasons.append("bloqueio de safetyRule crítico")
        # 2) alto risco declarado no objeto
        if self.cko.get("patientSafety", {}).get("highRisk"):
            for token in g.get("mandatoryHumanValidationFor", []):
                if token in ("high-risk", "high-risk-medication", "highRisk"):
                    needs_human = True
                    reasons.append("medicamento/dispositivo de alto risco")
        # 3) confiança abaixo do mínimo
        min_conf = sg.get("minConfidence", g.get("minOverallConfidence", 0.0))
        if self.trace.confidence is not None and self.trace.confidence < min_conf:
            needs_human = True
            reasons.append(f"confiança {self.trace.confidence} < mínimo {min_conf}")
        # 4) exigência explícita no estágio
        if sg.get("requiresHumanValidation"):
            needs_human = True
            reasons.append("requiresHumanValidation no estágio")

        if self.trace.safetyBlocks and any(sb for sb in self.trace.safetyBlocks):
            # never-event / bloqueio: não emite recomendação sem revisão
            self.trace.status = "blocked"
        elif needs_human:
            self.trace.status = "pending_human_validation"
        else:
            self.trace.status = "approved"
            self.trace.humanValidated = False  # aprovação automática permitida; sem humano ainda
        self.trace.reasoningTrace.append(
            {"stage": "humanValidation", "detail": f"status={self.trace.status}; motivos={reasons or ['nenhum']}"})

    def _stage_output(self, ctx: dict, stage: dict):
        self.trace.reasoningTrace.append({"stage": "output", "detail": "trace finalizado."})

    _HANDLERS = {
        "ingest": _stage_ingest,
        "factAssertion": _stage_fact_assertion,
        "ruleEvaluation": _stage_rule_evaluation,
        "inference": _stage_inference,
        "decision": _stage_decision,
        "workflow": _stage_workflow,
        "explanation": _stage_explanation,
        "humanValidation": _stage_human_validation,
        "output": _stage_output,
    }

    # ---- execução -------------------------------------------------------- #
    def run(self, patient_context: dict) -> ExecutionTrace:
        stages = self.layer.get("stages", [])
        if not stages:
            self.trace.status = "error"
            self.trace.warnings.append("ckosRuntimeLayer.stages ausente.")
            self.trace.completedAt = _now()
            return self.trace
        try:
            for st in stages:
                handler = self._HANDLERS.get(st["stage"])
                if handler is None:
                    self.trace.warnings.append(f"Estágio sem handler: {st['stage']}")
                    continue
                handler(self, patient_context, st)
                self.trace.stagesExecuted.append(st["stage"])
                # onFailure=halt em bloqueio de segurança
                if st.get("guardrails", {}).get("blockOnSafetyRule") and self.trace.safetyBlocks:
                    self.trace.status = "blocked"
                    self.trace.warnings.append(f"halt em {st['stage']} por bloqueio de segurança.")
                    break
        except Exception as exc:  # runtime robusto: erro não deve emitir recomendação insegura
            self.trace.status = "error"
            self.trace.warnings.append(f"Exceção em runtime: {exc}")
        self.trace.completedAt = _now()
        return self.trace


# --------------------------------------------------------------------------- #
# CLI
# --------------------------------------------------------------------------- #
def main():
    ap = argparse.ArgumentParser(description="Executa o pipeline CKOS sobre um documento CKO.")
    ap.add_argument("--schema")
    ap.add_argument("--cko", required=True)
    ap.add_argument("--patient", help="JSON inline do contexto do paciente.")
    ap.add_argument("--patient-file")
    args = ap.parse_args()

    cko = json.load(open(args.cko, encoding="utf-8"))
    if args.patient_file:
        patient = json.load(open(args.patient_file, encoding="utf-8"))
    elif args.patient:
        patient = json.loads(args.patient)
    else:
        patient = {}

    if args.schema:
        schema = json.load(open(args.schema, encoding="utf-8"))
        errs = CKOSRuntime.validate(cko, schema)
        print("== Validação estrutural ==")
        print("OK" if not errs else "\n".join(errs[:20]))
        print()

    trace = CKOSRuntime(cko).run(patient)
    print("== Execution Trace ==")
    print(json.dumps(trace.to_dict(), ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: cko-projeto/03-templates/ckos-runtime.js
/*!
 * ckos-runtime.js — CKOS Runtime v12 (browser port)
 * Porta o motor de referência Python (ckos_runtime.py) para o navegador e adiciona
 * dois itens do roadmap v12:
 *   (1) Parser de CKO-DSL: interpreta `governedRule.logic` (string) para conditions/actions.
 *   (2) Modelo de confiança ponderado (weighted), além do conservador (min).
 *
 * Sem dependências. Roda ao lado de cko-calc.js / cko-interactive.js e em preview.html.
 * Uso:
 *   import { CKOSRuntime, parseDSL } from './ckos-runtime.js';
 *   const trace = new CKOSRuntime(cko).run(patientContext);
 * ou via global:  window.CKOS.run(cko, patientContext)
 */
(function (root, factory) {
  const api = factory();
  if (typeof module === 'object' && module.exports) module.exports = api;
  root.CKOS = api;
}(typeof self !== 'undefined' ? self : this, function () {
  'use strict';

  // ----------------------------------------------------------------------- //
  // Operadores
  // ----------------------------------------------------------------------- //
  const OPS = {
    equals: (a, b) => a === b,
    notEquals: (a, b) => a !== b,
    greaterThan: (a, b) => a != null && a > b,
    lessThan: (a, b) => a != null && a < b,
    greaterOrEqual: (a, b) => a != null && a >= b,
    lessOrEqual: (a, b) => a != null && a <= b,
    in: (a, b) => Array.isArray(b) && b.includes(a),
    notIn: (a, b) => !(Array.isArray(b) && b.includes(a)),
    exists: (a, b) => (a != null) === !!b,
  };
  // Símbolos aceitos pela DSL → operador canônico
  const DSL_OPS = {
    '=': 'equals', '==': 'equals', '!=': 'notEquals', '<>': 'notEquals',
    '>': 'greaterThan', '<': 'lessThan', '>=': 'greaterOrEqual', '<=': 'lessOrEqual',
    'in': 'in', 'notin': 'notIn',
  };
  const ACTION_VERBS = { recommend: 'recommend', avoid: 'avoid', require: 'require', alert: 'alert', block: 'block' };

  // ----------------------------------------------------------------------- //
  // (1) Parser de CKO-DSL
  //   "IF medication = insulin AND volumeMl <= 0.3 THEN recommend seringa-x"
  //   Retorna { conditions:[{field,operator,value}], actions:[{type,target}] }
  // ----------------------------------------------------------------------- //
  function coerce(v) {
    if (v == null) return v;
    const t = v.trim();
    if (/^-?\d+(\.\d+)?$/.test(t)) return parseFloat(t);
    if (t === 'true') return true;
    if (t === 'false') return false;
    if (/^\[.*\]$/.test(t)) return t.slice(1, -1).split(',').map(s => coerce(s.trim()));
    return t.replace(/^["']|["']$/g, '');
  }

  function parseDSL(logic) {
    if (!logic || typeof logic !== 'string') return { conditions: [], actions: [] };
    const m = /^\s*IF\s+(.+?)\s+THEN\s+(.+?)\s*$/is.exec(logic);
    if (!m) return { conditions: [], actions: [], parseError: 'formato esperado: IF … THEN …' };

    const conditions = m[1].split(/\s+AND\s+/i).map(clause => {
      const cm = /^\s*(\S+)\s*(<=|>=|==|!=|<>|<|>|=|\bin\b|\bnotin\b)\s*(.+?)\s*$/i.exec(clause);
      if (!cm) return null;
      const operator = DSL_OPS[cm[2].toLowerCase()];
      return operator ? { field: cm[1], operator, value: coerce(cm[3]) } : null;
    }).filter(Boolean);

    const actions = m[2].split(/\s+AND\s+/i).map(clause => {
      const am = /^\s*(recommend|avoid|require|alert|block)\s+(.+?)\s*$/i.exec(clause);
      if (!am) return null;
      return { type: ACTION_VERBS[am[1].toLowerCase()], target: coerce(am[2]) };
    }).filter(Boolean);

    return { conditions, actions };
  }

  // Normaliza uma regra: usa conditions/actions se existirem; senão, faz parse do logic.
  function materialize(rule) {
    if ((rule.conditions && rule.conditions.length) || (rule.actions && rule.actions.length)) {
      return { conditions: rule.conditions || [], actions: rule.actions || [] };
    }
    return parseDSL(rule.logic);
  }

  // ----------------------------------------------------------------------- //
  // Avaliação
  // ----------------------------------------------------------------------- //
  const getField = (ctx, key) => {
    const v = ctx[key];
    return (v && typeof v === 'object' && 'value' in v) ? v.value : v;
  };
  const evalCondition = (c, ctx) => {
    const op = OPS[c.operator];
    if (!op) throw new Error('Operador desconhecido: ' + c.operator);
    return !!op(getField(ctx, c.field), c.value);
  };
  const evalRule = (conds, ctx) => conds.length > 0 && conds.every(c => evalCondition(c, ctx));

  // ----------------------------------------------------------------------- //
  // (2) Modelo de confiança
  // ----------------------------------------------------------------------- //
  function aggregateConfidence(supportRules, mode) {
    const confs = supportRules.map(r => (r.confidence != null ? r.confidence : 0.8));
    if (!confs.length) return 0;
    if (mode === 'min') return Math.min(...confs);
    // weighted: média ponderada por prioridade (prioridade menor = peso maior)
    let num = 0, den = 0;
    supportRules.forEach(r => {
      const w = 1 / Math.max(1, (r.priority != null ? r.priority : 100));
      const c = (r.confidence != null ? r.confidence : 0.8);
      num += c * w; den += w;
    });
    return den ? +(num / den).toFixed(3) : 0;
  }

  // ----------------------------------------------------------------------- //
  // Runtime
  // ----------------------------------------------------------------------- //
  class CKOSRuntime {
    constructor(cko, opts) {
      this.cko = cko || {};
      this.layer = this.cko.ckosRuntimeLayer || {};
      this.confidenceMode = (opts && opts.confidenceMode) || 'weighted';
      this.trace = {
        traceId: 'trace-' + Math.random().toString(36).slice(2),
        startedAt: new Date().toISOString(), completedAt: null,
        stagesExecuted: [], firedRules: [], assertedFacts: [], decisionPath: [],
        finalRecommendation: null, warnings: [], safetyBlocks: [],
        confidence: null, humanValidated: false, status: 'pending', reasoningTrace: [],
      };
    }

    _collectRules() {
      const repos = this.cko.ruleRepositories || {};
      const inline = (this.cko.clinicalKnowledge && this.cko.clinicalKnowledge.clinicalRules) || [];
      return [
        ['clinicalRules', (repos.clinicalRules || []).concat(inline)],
        ['safetyRules', repos.safetyRules || []],
        ['inventoryRules', repos.inventoryRules || []],
        ['educationRules', repos.educationRules || []],
      ];
    }

    _ingest(ctx) { this.trace.reasoningTrace.push({ stage: 'ingest', detail: Object.keys(ctx).length + ' campos de contexto.' }); }

    _factAssertion() {
      const a = (this.cko.factInferenceEngine && this.cko.factInferenceEngine.assertions) || [];
      a.forEach(x => this.trace.assertedFacts.push({ id: x.id, fact: x.fact, confidence: x.confidence != null ? x.confidence : 1 }));
      this.trace.reasoningTrace.push({ stage: 'factAssertion', detail: this.trace.assertedFacts.length + ' fatos assertidos.' });
    }

    _ruleEvaluation(ctx) {
      this._collectRules().forEach(([domain, rules]) => {
        rules.slice().sort((x, y) => (x.priority || 100) - (y.priority || 100)).forEach(r => {
          const { conditions, actions } = materialize(r);
          if (evalRule(conditions, ctx)) {
            const fired = {
              id: r.id || r.name || 'unnamed', domain: r.domain || domain,
              safetyCritical: !!r.safetyCritical, confidence: r.confidence != null ? r.confidence : 0.8,
              priority: r.priority != null ? r.priority : 100, actions,
            };
            this.trace.firedRules.push(fired);
            actions.forEach(act => {
              if ((act.type === 'avoid' || act.type === 'block') && fired.safetyCritical) {
                this.trace.safetyBlocks.push({ rule: fired.id, target: act.target, message: act.message || '' });
              }
            });
          }
        });
      });
      this.trace.reasoningTrace.push({ stage: 'ruleEvaluation', detail: this.trace.firedRules.length + ' regras dispararam; ' + this.trace.safetyBlocks.length + ' bloqueio(s).' });
    }

    _inference() {
      const known = new Set(this.trace.assertedFacts.map(f => f.id));
      const rules = (this.cko.factInferenceEngine && this.cko.factInferenceEngine.inferenceRules) || [];
      let changed = true;
      while (changed) {
        changed = false;
        rules.forEach(r => {
          const ant = r.if || [];
          if (ant.length && ant.every(x => known.has(x))) {
            (r.then || []).forEach(c => {
              if (!known.has(c)) { known.add(c); this.trace.assertedFacts.push({ id: c, fact: c, confidence: 0.9, inferred: true }); changed = true; }
            });
          }
        });
      }
      this.trace.reasoningTrace.push({ stage: 'inference', detail: this.trace.assertedFacts.filter(f => f.inferred).length + ' fato(s) inferido(s).' });
    }

    _decision(ctx) {
      const trees = (this.cko.decisionTreeRepository && this.cko.decisionTreeRepository.trees) || [];
      trees.forEach(tree => {
        const nodes = {}; (tree.nodes || []).forEach(n => nodes[n.id] = n);
        let cur = tree.rootNode;
        const guard = new Set();
        while (cur && nodes[cur] && !guard.has(cur)) {
          guard.add(cur);
          const node = nodes[cur];
          this.trace.decisionPath.push({ node: cur, question: node.question });
          if (node.type === 'terminal' || node.type === 'action') break;
          const ans = node.condition ? evalCondition(node.condition, ctx) : getField(ctx, node.field);
          cur = ans ? node.yes : node.no;
        }
      });
      const recs = this.trace.firedRules.flatMap(fr => fr.actions.filter(a => a.type === 'recommend' || a.type === 'require').map(a => ({ a, fr })));
      if (recs.length) {
        const support = this.trace.firedRules.filter(fr => fr.actions.some(a => a.type === 'recommend' || a.type === 'require'));
        this.trace.finalRecommendation = { recommendedDevice: recs[0].a.target, actionType: recs[0].a.type, supportingRules: support.map(fr => fr.id) };
        this._support = support;
      }
      this.trace.reasoningTrace.push({ stage: 'decision', detail: 'rec=' + JSON.stringify(this.trace.finalRecommendation) + '; caminho de ' + this.trace.decisionPath.length + ' nó(s).' });
    }

    _workflow() {
      const wfs = (this.cko.clinicalWorkflowEngine && this.cko.clinicalWorkflowEngine.workflows) || [];
      this.trace.reasoningTrace.push({ stage: 'workflow', detail: wfs.length + ' workflow(s).' });
    }

    _explanation() {
      const support = this._support || this.trace.firedRules;
      this.trace.confidence = aggregateConfidence(support, this.confidenceMode);
      this.trace.reasoningTrace.push({ stage: 'explanation', detail: 'confiança=' + this.trace.confidence + ' (modo ' + this.confidenceMode + ').' });
    }

    _humanValidation(stage) {
      const g = this.layer.globalGuardrails || {};
      const sg = (stage && stage.guardrails) || {};
      const reasons = []; let needsHuman = false;
      if (this.trace.safetyBlocks.length) { needsHuman = true; reasons.push('bloqueio de segurança'); }
      if (this.cko.patientSafety && this.cko.patientSafety.highRisk) {
        (g.mandatoryHumanValidationFor || []).forEach(t => {
          if (['high-risk', 'high-risk-medication', 'highRisk'].includes(t)) { needsHuman = true; reasons.push('alto risco'); }
        });
      }
      const minConf = sg.minConfidence != null ? sg.minConfidence : (g.minOverallConfidence || 0);
      if (this.trace.confidence != null && this.trace.confidence < minConf) { needsHuman = true; reasons.push('confiança ' + this.trace.confidence + ' < ' + minConf); }
      if (sg.requiresHumanValidation) { needsHuman = true; reasons.push('requiresHumanValidation'); }

      if (this.trace.safetyBlocks.length) this.trace.status = 'blocked';
      else if (needsHuman) this.trace.status = 'pending_human_validation';
      else { this.trace.status = 'approved'; this.trace.humanValidated = false; }
      this.trace.reasoningTrace.push({ stage: 'humanValidation', detail: 'status=' + this.trace.status + '; motivos=' + JSON.stringify(reasons.length ? reasons : ['nenhum']) });
    }

    _output() { this.trace.reasoningTrace.push({ stage: 'output', detail: 'trace finalizado.' }); }

    run(patientContext) {
      const ctx = patientContext || {};
      const H = {
        ingest: () => this._ingest(ctx), factAssertion: () => this._factAssertion(),
        ruleEvaluation: () => this._ruleEvaluation(ctx), inference: () => this._inference(),
        decision: () => this._decision(ctx), workflow: () => this._workflow(),
        explanation: () => this._explanation(), humanValidation: (st) => this._humanValidation(st),
        output: () => this._output(),
      };
      const stages = this.layer.stages || [];
      if (!stages.length) { this.trace.status = 'error'; this.trace.warnings.push('ckosRuntimeLayer.stages ausente.'); this.trace.completedAt = new Date().toISOString(); return this.trace; }
      try {
        for (const st of stages) {
          const h = H[st.stage];
          if (!h) { this.trace.warnings.push('sem handler: ' + st.stage); continue; }
          h(st); this.trace.stagesExecuted.push(st.stage);
          if (st.guardrails && st.guardrails.blockOnSafetyRule && this.trace.safetyBlocks.length) {
            this.trace.status = 'blocked'; this.trace.warnings.push('halt em ' + st.stage + ' por bloqueio de segurança.'); break;
          }
        }
      } catch (e) { this.trace.status = 'error'; this.trace.warnings.push('exceção: ' + e.message); }
      this.trace.completedAt = new Date().toISOString();
      return this.trace;
    }
  }

  return {
    CKOSRuntime, parseDSL, materialize, aggregateConfidence,
    run: (cko, ctx, opts) => new CKOSRuntime(cko, opts).run(ctx),
    version: '12.0.0',
  };
}));
